lib/twitter/twits.ex

```diff
@@ -0,0 +1,163 @@
+  """
+  def get_post!(%Accounts.User{} = user, id) do
+    Post
+    |> user_posts_query(user)
+    |> Repo.get!(id)
+  end
+
+  @spec create_post(User.t(), map()) :: {:ok, Post.t()} | {:error, Ecto.Changeset.t(Post.t())}
+  @doc """
+  Creates a post. current_user id is registered automatically
+  ## Examples
+      iex> create_post(%{field: value})
+      {:ok, %Post{}}
+      iex> create_post(%{field: bad_value})
+      {:error, %Ecto.Changeset{}}
+  """
+  def create_post(%Accounts.User{} = user, attrs \\ %{}) do
+    %Post{}
+    |> Post.changeset(attrs)
+    |> Ecto.Changeset.put_assoc(:user, user)
+    |> Repo.insert()
+  end
+
+  @spec update_post(Post.t(), map()) :: {:ok, Post.t()} | {:error, Ecto.Changeset.t(Post.t())}
+  @doc """
+  Updates a post.
+  ## Examples
+      iex> update_post(post, %{field: new_value})
+      {:ok, %Post{}}
+      iex> update_post(post, %{field: bad_value})
+      {:error, %Ecto.Changeset{}}
+  """
+  def update_post(%Post{} = post, attrs) do
+    post
+    |> Post.changeset(attrs)
+    |> Repo.update()
+  end
+
+  @spec delete_post(Post.t()) :: {:ok, Post.t()} | {:error, Ecto.Changeset.t(Post.t())}
+  @doc """
+  Deletes a post.
+  ## Examples
+      iex> delete_post(post)
+      {:ok, %Post{}}
+      iex> delete_post(post)
+      {:error, %Ecto.Changeset{}}
+  """
+  def delete_post(%Post{} = post) do
+    Repo.delete(post)
+  end
+
+  @spec change_post(Post.t(), map()) :: Ecto.Changeset.t(Post.t())
+  @doc """
+  Returns an `%Ecto.Changeset{}` for tracking post changes.
+  ## Examples
+      iex> change_post(post)
+      %Ecto.Changeset{data: %Post{}}
+  """
+  def change_post(%Post{} = post, attrs \\ %{}) do
+    Post.changeset(post, attrs)
+  end
+
+  ################################################
+  ############# LIKES ###########################
+  ################################################
+  defp likes_of_post_query(query, post_id) do
+    from l in query, where: l.post_id == ^post_id
+  end
+
+  defp get_like_query(query, post_id, %Accounts.User{id: user_id}) do
+    from l in query, where: l.post_id == ^post_id and l.user_id == ^user_id
+  end
+
+  @spec list_likes_by_post_id(integer() | String.t()) :: list(Like.t())
+  @doc """
+  Returns the likes of a post.
+  """
+  def list_likes_by_post_id(post_id) do
+    Like
+    |> likes_of_post_query(post_id)
+    |> Repo.all()
+  end
+
+  @doc """
+  like someone else's post: the post_id and current user should be passed
+  current user will be assigned automatically with actions in like_controller
+  """
+  @spec create_like(User.t(), integer() | String.t()) :: Like.t()
+  def create_like(%Accounts.User{} = user, post_id) do
+    # check that it doesn't exist
+    query = get_like_query(Like, post_id, user)
+
+    with false <- Repo.exists?(query) do
+      %Like{}
+      |> Like.changeset(%{user_id: user.id, post_id: post_id})
+      # |> Ecto.Changeset.put_assoc(:post, post)
+      |> Repo.insert()
+    end
+  end
```

Comment:
no need to check that it doesn't exist. Just attempt to insert it, let the database give you a unique constraint error and handle that in the controller.

Reply:
Ok, fixed that